src/som/interpreter/actors/Actor.java

```diff
@@ -190,12 +203,73 @@ protected void appendToMailbox(final EventualMessage msg) {
     mailboxExtension.append(msg);
   }
 
-  protected static void handleBreakPoints(final EventualMessage msg, final WebDebugger dbg) {
-    if (VmSettings.TRUFFLE_DEBUGGER_ENABLED && msg.isMessageReceiverBreakpointSet()) {
+  protected static void handleBreakpointsAndStepping(final EventualMessage msg, final WebDebugger dbg, final Actor actor) {
+    if (!VmSettings.TRUFFLE_DEBUGGER_ENABLED) {
+      return;
+    }
+
+    synchronized (actor) {
+
+      if (msg.isStepToReceiver() || msg.isMessageReceiverBreakpointSet() || actor.isStepToNextTurn()) {
+        dbg.prepareSteppingUntilNextRootNode();
+        if (actor.isStepToNextTurn()) { // reset flag
+          actor.setStepToNextTurn(false);
+        }
+      }
+
+      if (msg.isStepToNextTurn()) {
+        prepareStepToNextMessage(msg, actor);
+      }
+
+      if (msg.isStepToPromiseResolution()) {
+        prepareStepToResolution(msg);
+      }
+
+      if (msg.isStepReturnFromTurnToPromise()) {
+        prepareStepToResolution(msg.getCausalMessage());
+      }
+
+      prepareCallbackForStepping(msg, dbg);
+  }
+
+ }
+
+  /**
+   * Check callbacks for stop in the resolution of the promise.
+   */
+  private static void prepareCallbackForStepping(final EventualMessage msg, final WebDebugger dbg) {
+    boolean triggerStepping = false;
+    if (msg instanceof PromiseSendMessage) {
+      if (((PromiseSendMessage) msg).getPromise().isTriggerStopBeforeExecuteCallback()) {
```

Comment:
Because in the `PromiseSendMessage` the method `getPromise` returns the promise to which that message was sent. In the case of `PromiseCallbackMessage` the method `getPromise` refers to the promise that is returned by the whenResolved. If we want to handle both cases with one approach, a solution will be to swap the semantics of `getPromise` and `getPromiseRegisteredOn` in `PromiseCallbackMessage`, then we could use `PromiseMessage.getPromise` method instead of the two separated approaches.